### ISSUE TITLE
zc156: Incompatible changes

### ISSUE THREAD

**Issue by lat9** [zencart/zencart#2162]
Zen Cart's strength, IMO, is the wide swath of add-ons/plugins available for a store-owner to customize his/her store.  Unfortunately, each Zen Cart release seems to throw multiple, incompatible monkey-wrenches into the mix.

Some of these incompatible changes are necessary, e.g. the `admin_pages` support introduced in zc150 and the admin-sanitization introduced in zc155, but some aren't.

-----

To wit, [this](https://github.com/zencart/zencart/commit/aa5426d453f36ebbd62e6354c63e7b1c45db3301#diff-decb27dc96b8dede41256870a356197b) zc156 commit which modifies the loading order of the admin-level language files &mdash; which had been otherwise unchanged since zc138a (and earlier). So far, and I believe there will be others, the [Sales Report](https://www.zen-cart.com/downloads.php?do=file&id=9) plugin has been victimized by this change. That plugin's language file makes (or, now, _made_) use of a base language file's `DATE_FORMAT` to determine the format in which to display various dates in the report. The change associated with the referenced commit has, essentially, broken a plugin that has been operating (bug-fixes excepted) since zc150.

-----

On another vein, **notifiers**, which allow plugins to seamlessly add their required changes without carving up Zen Cart core files.  When zc156 _chose_ to re-format the admin's ***Categories / Products*** listing, the behavior associated with two notifications introduced in zc155 was _silently_ changed.

In zc155, the notification issued by `/admin/categories.php`

```
  $zco_notifier->notify('NOTIFY_BEGIN_ADMIN_CATEGORIES', $action);
```

served to notify that a category-related action was in progress. The notification issued by `/admin/product.php`

```
  $zco_notifier->notify('NOTIFY_BEGIN_ADMIN_PRODUCTS', $action);
```

provided similar notification on a product level.

The change in zc156 migrated the handling of many of those category/product actions to `/admin/category_products_listing.php` but failed to carry over the associated notifications, leaving any plugin that tried to "do the right thing" to silently fail in its attempts.

-------

In summary, Zen Cart releases need to pay more attention to maintaining a compatible interface for plugin developers, especially when a change is either a solution-looking-for-a-problem or cosmetic.  As a plugin developer, I don't mind making updates associated with necessary changes, but it drives me nuts to have to fiddle with working plugins to provide a work-around for those optional changes.

**Comment by drbyte** [comment]
Noted. There were a few changes missed during the code review process where there were massive changes to some files, inside which these unexpected changes were buried.

While we try to test and verify as much as possible, sometimes things get missed. The source is open, so community testing is invited and helpful. Clearly these didn't get tested by people using the affected addons.
